src/main/java/akka/stream/contrib/FileTailSource.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib;
+
+import akka.NotUsed;
+import akka.stream.Attributes;
+import akka.stream.Outlet;
+import akka.stream.SourceShape;
+import akka.stream.javadsl.Source;
+import akka.stream.stage.*;
+import akka.util.ByteString;
+import scala.concurrent.duration.FiniteDuration;
+import scala.util.Failure;
+import scala.util.Success;
+import scala.util.Try;
+
+import java.nio.ByteBuffer;
+import java.nio.channels.AsynchronousFileChannel;
+import java.nio.channels.CompletionHandler;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+
+/**
+ * Read the entire contents of a file, and then when the end is reached, keep reading
+ * newly appended data. Like the unix command `tail -f`.
+ *
+ * Aborting the stage can be done by combining with a [[akka.stream.KillSwitch]]
+ */
+public final class FileTailSource extends GraphStage<SourceShape<ByteString>> {
+
+  private final Path path;
+  private final int maxChunkSize;
+  private final long startingPosition;
+  private final FiniteDuration pollingInterval;
+  private final Outlet<ByteString> out = Outlet.create("FileTailSource.out");
+  private final SourceShape<ByteString> shape = SourceShape.of(out);
+
+  // this is stateless, so can be shared among instances
+  private static final CompletionHandler<Integer, AsyncCallback<Try<Integer>>> completionHandler = new CompletionHandler<Integer, AsyncCallback<Try<Integer>>>() {
+    @Override
+    public void completed(Integer result, AsyncCallback<Try<Integer>> attachment) {
+      attachment.invoke(new Success<>(result));
+    }
+
+    @Override
+    public void failed(Throwable exc, AsyncCallback<Try<Integer>> attachment) {
+      attachment.invoke(new Failure<>(exc));
+    }
+  };
+
+  public FileTailSource(Path path, int maxChunkSize, long startingPosition, FiniteDuration pollingInterval) {
+    this.path = path;
+    this.maxChunkSize = maxChunkSize;
+    this.startingPosition = startingPosition;
+    this.pollingInterval = pollingInterval;
+  }
+
+  @Override
+  public SourceShape<ByteString> shape() {
+    return shape;
+  }
+
+  @Override
+  public GraphStageLogic createLogic(Attributes inheritedAttributes) {
+    try {
+      if (!Files.exists(path)) throw new IllegalArgumentException("Path '" + path + "' does not exist");
+      if (Files.isDirectory(path)) throw new IllegalArgumentException("Path '" + path + "' cannot be tailed, it is a directory");
+      if (!Files.isReadable(path)) throw new IllegalArgumentException("No read permission for '" + path + "'");
+
+      return new TimerGraphStageLogic(shape) {
+        private final ByteBuffer buffer = ByteBuffer.allocate(maxChunkSize);
+        private final AsynchronousFileChannel channel = AsynchronousFileChannel.open(path, StandardOpenOption.READ);
```

Comment:
This runs the IO on "the default thread pool", not sure what kind of thread pool that is or if it is something to worry about.